specifyweb/frontend/js_src/lib/components/Attachments/RecordSetAttachment.tsx

```diff
@@ -0,0 +1,113 @@
+import { AttachmentGallery } from './Gallery';
+
+export function RecordSetAttachments<SCHEMA extends AnySchema>({
+  records,
+  onClose: handleClose,
+  onFetch: handleFetch,
+}: {
+  readonly records: RA<SpecifyResource<SCHEMA> | undefined>;
+  readonly onClose: () => void;
+  readonly onFetch?:
+    | ((index: number) => Promise<RA<number | undefined> | void>)
+    | undefined;
+}): JSX.Element {
+  const recordFetched = React.useRef<number>(0);
+
+  const [attachments] = useAsyncState(
+    React.useCallback(async () => {
+      const relatedAttachementRecords = await Promise.all(
+        records.map((record) =>
+          record
+            ?.rgetCollection(`${record.specifyModel.name}Attachments`)
+            .then(
+              ({ models }) =>
+                models as RA<SpecifyResource<CollectionObjectAttachment>>
+            )
+        )
+      );
+
+      const fetchCount = records.findIndex(
+        (record) => record?.populated !== true
+      );
+
+      recordFetched.current = fetchCount === -1 ? records.length : fetchCount;
+
+      return Promise.all(
+        filterArray(relatedAttachementRecords.flat()).map(
+          async (collectionObjectAttachment) => ({
+            attachment: await collectionObjectAttachment
+              .rgetPromise('attachment')
+              .then((resource) => serializeResource(resource)),
+            related: collectionObjectAttachment,
+          })
+        )
+      );
+    }, [records]),
+    true
+  );
+
+  const [haltValue, setHaltValue] = React.useState(300);
+  const halt = attachments?.length === 0 && records.length >= haltValue;
+
+  const [scale = defaultScale] = useCachedState('attachments', 'scale');
+
+  const children = halt ? (
+    haltValue === records.length ? (
+      <>{attachmentsText.noAttachments()}</>
+    ) : (
+      <div className="flex flex-col gap-4">
+        {attachmentsText.attachmentHaltLimit({ halt: haltValue })}
+        <Button.Orange
+          onClick={() => {
+            if (haltValue + 300 > records.length) {
+              setHaltValue(records.length);
+            } else {
+              setHaltValue(haltValue + 300);
+            }
+          }}
+        >
+          {attachmentsText.fetchNextAttachments()}
+        </Button.Orange>
+      </div>
+    )
+  ) : (
+    <AttachmentGallery
+      attachments={attachments?.map(({ attachment }) => attachment) ?? []}
+      isComplete={recordFetched.current === records.length}
+      scale={scale}
+      onChange={(attachment, index): void =>
+        void attachments?.[index].related.set(`attachment`, attachment)
+      }
+      onFetchMore={
+        attachments === undefined || handleFetch === undefined || halt
+          ? undefined
+          : async () => handleFetch?.(recordFetched.current)
+      }
+    />
+  );
+
+  return (
+    <Dialog
+      buttons={<Button.DialogClose>{commonText.close()}</Button.DialogClose>}
+      header={attachmentsText.attachments()}
```

Comment:
feature: display attachment count in here. think of how to indicate that not all attachments might have been fetched yet. or if you want to keep it simple just display the `records.length`, but only once all attachments are fetched.

consider using `commonText.countLine()` for this (see other usages in the code)